src/include/module/module/generic.h

```diff
@@ -0,0 +1,138 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Copyright(c) 2023 Intel Corporation. All rights reserved.
+ *
+ * Author: Marcin Rajwa <marcin.rajwa@linux.intel.com>
+ *	   Adrian Warecki <adrian.warecki@intel.com>
+ */
+
+#ifndef __MODULE_MODULE_GENERIC__
+#define __MODULE_MODULE_GENERIC__
+
+#include <stdint.h>
+#include <stddef.h>
+#include <stdbool.h>
+
+#include "interface.h"
+#include "../ipc4/base-config.h"
+
+#define module_get_private_data(mod) ((mod)->priv.private)
+#define module_set_private_data(mod, data) ((mod)->priv.private = data)
+
+/**
+ * \struct module_config
+ * \brief Module config container, used for both config types.
+ */
+struct module_config {
+	size_t size; /**< Specifies the size of whole config */
+	bool avail; /**< Marks config as available to use.*/
+	void *data; /**< tlv config, a pointer to memory where config is stored. */
+	const void *init_data; /**< Initial IPC configuration. */
+#if CONFIG_IPC_MAJOR_4
+	struct ipc4_base_module_cfg base_cfg;
+#endif
+};
+
+/** private, runtime module data */
+struct module_data {
+	void *private; /**< self object, memory tables etc here */
+	struct module_config cfg; /**< module configuration data */
+#ifdef SOF_MODULE_API_PRIVATE
+	enum module_state state;
+	size_t new_cfg_size; /**< size of new module config data */
+	void *runtime_params;
+	const struct module_interface *ops; /**< module specific operations */
+	struct module_memory memory; /**< memory allocated by module */
+	struct module_processing_data mpd; /**< shared data comp <-> module */
+	void *module_adapter; /**<loadable module interface handle */
+	uint32_t module_entry_point; /**<loadable module entry point address */
+#endif /* SOF_MODULE_PRIVATE */
+};
+
+/* module_adapter private, runtime data - public version*/
+struct processing_module {
+	struct module_data priv; /**< module private data */
+	uint32_t period_bytes; /** pipeline period bytes */
+#ifdef SOF_MODULE_API_PRIVATE
```

Comment:
is this macro only going to be defined in the base firmware and not in modules? I had problems with conflicting structure definitions when failing to keep them in sync, so this can be dangerous